pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,494 @@
+// - finds the most optimal candidate and elect
+// Instead of keeping a map and lock on election, the state is
+// calculated every time by looking at the lease, and set of available candidates.
+// PingTime + 5 seconds > time.Now: We just asked all candidates to ack and are still waiting for response
+// RenewTime + 5 seconds > time.Now: All candidates acked in the last 5 seconds, proceed with election
+// The missing case is that the election should auto proceed after 5 seconds even if we don't get all acks.
+// I think we should be able to find the right state with the information available...
+func (c *Controller) stepElection(ctx context.Context, leaderLeaseID leaderLeaseID) error {
+	now := time.Now()
+	if leaderLeaseID.name == controllerName {
+		return fmt.Errorf("lease not managed by CLE")
+	}
+
+	klog.Infof("stepElection %q %q", leaderLeaseID.namespace, leaderLeaseID.name)
+
+	candidates, err := c.listAdmissableCandidates(leaderLeaseID)
+	if err != nil {
+		return err
+	}
+
+	// Check if an election is really needed by looking at the current lease
+	// and set of candidates
+	reelection, err := c.electionNeeded(candidates, leaderLeaseID)
+	if err != nil {
+		return err
+	} else if !reelection {
+		return nil
+	}
+
+	for _, candidate := range candidates {
+		// If PingTime is within last 5 seconds, some candidates still have not acked
+		// and we should not proceed
+		if candidate.Spec.PingTime != nil && candidate.Spec.PingTime.Add(5*time.Second).Before(now) {
+			return nil
+		}
+	}
+
+	continueElection := true
+	for _, candidate := range candidates {
+		// if renewTime of ALL candidates is within last 5 seconds, continue election
+		// In other words, if renewTime of a single candidate is outside of 5 seconds, we need to resend
+		if candidate.Spec.RenewTime != nil && candidate.Spec.RenewTime.Add(5*time.Second).Before(now) {
+			continueElection = false
+			break
+		}
+	}
+	if !continueElection {
+		// Send an "are you alive" signal to all candidates
+		for _, candidate := range candidates {
+			clone := candidate.DeepCopy()
+			clone.Spec.PingTime = &metav1.MicroTime{Time: time.Now()}
+			_, err := c.leaseCandidateClient.LeaseCandidates(clone.Namespace).Update(ctx, clone, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		c.electionQueue.AddAfter(leaderLeaseID, 4*time.Second)
+		return nil
+	}
+
+	var ackedCandidates []*v1alpha1.LeaseCandidate
+	for _, candidate := range candidates {
+		if candidate.Spec.RenewTime.Add(5 * time.Second).After(now) {
+			ackedCandidates = append(ackedCandidates, candidate)
+		}
+	}
+	if len(ackedCandidates) == 0 {
+		return fmt.Errorf("no available candidates")
+	}
+
+	strategy := pickBestStrategy(ackedCandidates)
+	if strategy != v1.OldestEmulationVersion {
+		return fmt.Errorf("strategy %s is not recognized by CLE", strategy)
+	}
+
+	electee := pickBestLeader(ackedCandidates)
+
+	if electee == nil {
+		return fmt.Errorf("should not happen, could not find suitable electee")
```

Comment:
do we want to retry later?

Reply:
electee is calculated solely based on the state of leasecandidates. If we have the exact same state, we will reach the same conclusion with a retry. IMO we should not retry, but any changes to state will trigger the informer eventhandler to attempt the election again.